File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet (..), ageOn) where

data Planet
  = Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Uranus
  | Neptune

earthYear :: Float
earthYear = 31557600

ageOn :: Planet -> Float -> Float
ageOn = calculateAge . byYear
 where
  byYear p = case p of
    Mercury -> 0.2408467 * earthYear
    Venus -> 0.61519726 * earthYear
    Earth -> earthYear
    Mars -> 1.8808158 * earthYear
    Jupiter -> 11.862615 * earthYear
    Saturn -> 29.447498 * earthYear
    Uranus -> 84.016846 * earthYear
    Neptune -> 164.79132 * earthYear

  calculateAge y s =
    (/ 100) . fromIntegral . round . (* 100) $ s / y

File: zig/space-age/space_age.zig
pub const Planet = enum {
    mercury,
    venus,
    earth,
    mars,
    jupiter,
    saturn,
    uranus,
    neptune,

    fn getOrbitalPeriod(self: Planet) f64 {
        return switch (self) {
            .mercury => 0.2408467,
            .venus => 0.61519726,
            .earth => 1.0,
            .mars => 1.8808158,
            .jupiter => 11.862615,
            .saturn => 29.447498,
            .uranus => 84.016846,
            .neptune => 164.79132,
        };
    }

    pub fn age(self: Planet, seconds: usize) f64 {
        const planetYear = 31557600.0 * getOrbitalPeriod(self);
        return @intToFloat(f64, seconds) / planetYear;
    }
};
